cli/tests/cli_test.go

```diff
@@ -0,0 +1,245 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tests
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"testing"
+)
+
+var (
+	cwd       string
+	cftBinary string
+)
+
+// TestCLI does integration testing.
+func TestCLI(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping integration tests in short mode.")
+	}
+
+	// Test cases for each command.
+	cases := []struct {
+		name    string
+		testers []struct {
+			name   string
+			tester func(t *testing.T, dirTestdata string)
+		}
+		setup   func(t *testing.T, dirTestdata string)
+		destroy func(t *testing.T, dirTestdata string)
+	}{
+		{
+			name: "validate",
+			testers: []struct {
+				name   string
+				tester func(t *testing.T, dirTestdata string)
+			}{
+				{
+					name:   "passing",
+					tester: testValidatePassing,
```

Comment:
As mentioned, these test cases and functions seem a bit overly generic. I'd try to make this much closer to the tf-validator tests.

In fact, we can probably mirror those almost exactly - https://github.com/GoogleCloudPlatform/terraform-validator/blob/master/test/cli_test.go#L49

Reply:
No really, because logic and lifecycle is different.

For example, DM requires real resources to be created (even in preview mode we need to create previews)